update dynamic metrics article to work with case study (#67)
tidymodels/tidymodels.org: learn/statistics/survival-metrics/index.qmd

#| label: data
library(tidymodels)
library(censored)

> data(building_complaints, package = "modeldatatoo")

Reviewer: ```suggestion
building_complaints <- modeldatatoo::data_building_complaints()
```

building_complaints <- building_complaints %>% 
  mutate(
    disposition_surv = Surv(days_to_disposition, status == "CLOSED"), 
    .keep = "unused"
  )

set.seed(403)
complaints_split <- initial_validation_split(building_complaints)
complaints_train <- training(complaints_split)
complaints_val <- validation(complaints_split)
```

> We'll need a model to illustrate the code and concepts. Let's fit a basic proportional hazards model to the training set. We'll do a little bit of work on some of the predictors that have many possible levels using a recipe:

Reviewer: ```suggestion
We'll need a model to illustrate the code and concepts. Let's fit a basic Weibull model to the training set. We'll do a little bit of work on some of the predictors that have many possible levels using a recipe:
```

```{r}

> #| label: bag-tree-fit

Reviewer: I would rename objects in this chunk to `complaints_*` (spec, recipe, wflow, fit).

where `eval_time` is a vector of time points at which we want the corresponding survivor function estimates. Alternatively, we can use the `augment()` function to get both types of prediction and automatically attach them to the data. 

> We’ll use a finer grid than the original analysis with a maximum evaluation time of 200 days for this analysis. 

Reviewer: ```suggestion
We’ll use a finer grid than the [case study](../survival-case-study) with a maximum evaluation time of 200 days for this analysis. 
```

```{r}
#| label: val-pred

> time_points <- seq(0, 200, by = 10) # 21 evaluation times

Reviewer: ```suggestion
time_points <- seq(0, 200, by = 10)
```
slight preference for removing the comment since it's mentioned below but also not mad if the comment stays

## Brier Score

The Brier score is a metric that can be used with both classification and event-time models. In classification models, we compute the squared error between the observed outcome (encoded as 0/1) and the corresponding predicted probability for the class. 

A little math: suppose that the value $y_{ik}$ is a 0/1 indicator for whether the observed outcome $i$ corresponds to class $k$, and $\hat{p}_{ik}$ is the estimated class probability. The classification score is then:

$$
Brier_{class} = \frac{1}{N}\sum_{i=1}^N\sum_{k=1}^C (y_{ik} - \hat{p}_{ik})^2
$$

> For survival models, we transform the event time data into a binary version $y_{it}$: is there an event at evaluation time $t$^[Again, see the [Accounting for Censoring in Performance Metrics for Event Time Data](../survival-metrics-details) article for more on this.]. The survival function estimates $\hat{p}_{it}$ is the probability corresponding to non-events at time $t$. For example, if there has not been an event at the current evaluation time, our best model should estimate the survival probability near one. For observations that are events, the probability estimate is just one minus the survivor estimate. To account for censoring, we also weight each observation with $w_{it}$. The [time-dependent Brier score](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C7&q=%22Assessment+and+Comparison+of+Prognostic+Classification+Schemes+for+Survival+Data.%22&btnG=) is: 

Reviewer: ```suggestion
For survival models, we transform the event time data into a binary version $y_{it}$: is there an event at evaluation time $t$^[Again, see the [Accounting for Censoring in Performance Metrics for Event Time Data](../survival-metrics-details) article for more on this.]. The survival function estimate $\hat{p}_{it}$ is the probability corresponding to non-events at time $t$. For example, if there has not been an event at the current evaluation time, our best model should estimate the survival probability near one. For observations that are events, the probability estimate is just one minus the survivor estimate. To account for censoring, we also weight each observation with $w_{it}$. The [time-dependent Brier score](https://scholar.google.com/scholar?hl=en&as_sdt=0%2C7&q=%22Assessment+and+Comparison+of+Prognostic+Classification+Schemes+for+Survival+Data.%22&btnG=) is: 
```

  labs(x = "time", y = "Brier score")
```

> This shows the worst predictions (relatively speaking) occur at `r brier_scores %>% slice_max(.estimate, n = 1) %>% pluck(".eval_time")` days with a corresponding Brier score of `r brier_scores %>% slice_max(.estimate, n = 1) %>% pluck(".estimate") %>% signif(digits = 3)` Performance gets steadily better over (evaluation) time. 

Reviewer: ```suggestion
This shows the worst predictions (relatively speaking) occur at `r brier_scores %>% slice_max(.estimate, n = 1) %>% pluck(".eval_time")` days with a corresponding Brier score of `r brier_scores %>% slice_max(.estimate, n = 1) %>% pluck(".estimate") %>% signif(digits = 3)`. Performance gets steadily better over (evaluation) time. 
```

Reviewer: This here suggests to me that calibration is fairly good for higher eval times. See comment on line 519 for a possible contradiction.

## Receiver Operating Characteristic (ROC) Curves

When we not only turn the event time data into a binary representation but also the predicted probabilities, we are in well-chartered classification metrics territory. Sensitivity and specificity are common quantities to compute; we do so here in their weighted version to account for censoring:

- Sensitivity: How well do we predict the events? This is analogous to the true positive rate.
- Specificity: How well do we predict the non-events? One minus specificity is the false positive rate. 

These depend on the threshold used to turn predicted probabilities into predicted events/non-events. Let's look at the distribution of the survival probabilities for our example data at an evaluation time of 10 days. The distributions are separated by the observed class and weighted by the censoring weights. Details of both aspects are the same as the Brier score and can be found in the [Accounting for Censoring in Performance Metrics for Event Time Data](../survival-metrics-details) article.

```{r}
#| label: early-data
#| include: false 
time_as_binary_event <- function(surv, eval_time) {

>   disposition_surv <- .extract_surv_time(surv)

Reviewer: Was this a find-and-replace-all or did you intentionally rename this inside this function here? While I appreciate the link to the dataset at hand, I would not make this change. If you want to link it closer, you could rename `surv` to `disposition_surv` but I would prefer this to have general names since it's a function.

}

# Unnest the list columns and convert the event time data to binary format 
binary_encoding <- 
  val_pred %>% 
  select(.pred, disposition_surv) %>% 
  add_rowindex() %>% 
  unnest(.pred) %>% 
  mutate(
    obs_class = time_as_binary_event(disposition_surv, .eval_time),
    pred_class = if_else(.pred_survival >= 1 / 2, "non-event", "event"),
    pred_class = factor(pred_class, levels = c("event", "non-event")),
    .pred_event = 1 - .pred_survival
  )

early_data <- 
  binary_encoding %>% 
  filter(.eval_time == 10 & !is.na(.weight_censored)) %>% 
  select(.eval_time, .pred_survival, .weight_censored, obs_class, pred_class, 
         .pred_event, disposition_surv)

if (!file.exists("binary_encoding.RData")) {
  save(binary_encoding, file = "binary_encoding.RData")
}
```

```{r}
#| label: surv-hist-early
#| echo: false
#| warning: false
#| out-width: 70%
#| fig-width: 7
#| fig-height: 7
early_data %>% 
  ggplot(aes(x = 1 - .pred_survival, weight = .weight_censored)) + 
  geom_vline(xintercept = 1 / 2, col = "blue", lty = 2) +
  geom_histogram(col = "white", bins = 30) + 
  facet_wrap(~obs_class, ncol = 1) +
  lims(x = 0:1) +

>   labs(x = "probability of event at or before 10 weeks\n(1 - probability of 10 week survival)", y = "sum of weights") +

Reviewer: Currently all the labels say "probability" although they are predicted probabilities. I think it would be worth it to clarify that on all of them.

           estimate = pred_class,
           case_weights = .weight_censored)

pred_early <- augment(survreg_fit, complaints_val, eval_time = 10)

curve_early <- pred_early %>% roc_curve_survival(truth = disposition_surv, .pred)
auc_early <- pred_early %>% roc_auc_survival(truth = disposition_surv, .pred)
```

> More probability values are to the right of the 50% cutoff for the true non-events. Conversely, true events tend to have smaller probabilities. Using this cutoff, the sensitivity would be `r round(stats_early$.estimate[2] * 100, 1)`% and the specificity would be `r round(stats_early$.estimate[3] * 100, 1)`%. There are other possible cutoffs for the survival probabilities. Maybe one of these would have better statistics. 

Reviewer: ```suggestion
More probability values are to the left of the 50% cutoff for the true non-events. The true events tend to have larger probabilities but are not nearly as clearly distributed to the right of the cutoff as the non-events are to the left of the cutoff. Using this cutoff, the sensitivity would be `r round(stats_early$.estimate[2] * 100, 1)`% and the specificity would be `r round(stats_early$.estimate[3] * 100, 1)`%. There are other possible cutoffs for the survival probabilities. Maybe one of these would have better statistics. 
```

  labs(x = "time", y = "ROC AUC")
```

In this case, performance is best at earlier time points (unlike the Brier score), degrades a bit, and then increases again. Despite this, performance is fairly good across all non-zero evaluation times. 

## Disagreement between metrics

> It should not be surprising that each metric's results differ over time, but it may seem odd that the best model results differ between metrics. 

Reviewer: ```suggestion
While it may not be surprising that each metric's results vary over time, it may seem odd that _when_ the model does best also differs between metrics. 
```

The issue is that the ROC measures class separation, and the Brier score focuses more on accurate and well-calibrated predictions. These are not the same thing. As we'll see shortly, it can be easy to separate data between qualitative prediction (event or no event) even when the corresponding probability predictions are very inaccurate. 

```{r}
#| label: example-miscal-data
#| echo: false
#| fig-height: 4
#| fig-width: 8
sim_n <- 500
example_data <- 
  tibble(
    .pred_event = c(rnorm(sim_n, mean = .55, sd = 0.025),
                    rnorm(sim_n, mean = .45, sd = 0.025)),
    class = factor(rep(c("event", "no-event"), each = sim_n))
  )

p_ex_dens <- 
  example_data %>% 
  ggplot(aes(.pred_event, col = class, fill = class)) + 

>   stat_density(alpha = 1 / 4, trim = TRUE, adjust = 1.2) + 

Reviewer: A histogram would make more sense to me here. That matches what we had before. With a density, I'm tempted to think this is the underlying true distribution but that's not it here.

Author: I really dislike overlapping histograms. I've changed the label to be `"predicted probability...` to avoid ambiguity

    
    dat <- get_binary_data(eval_time)
    
    cal_data(
      dat$obs_class,
      dat$.pred_event,
      dat$.weight_censored,
      cuts$lower_cut,
      cuts$mid,
      cuts$upper_cut
    ) %>%
      mutate(.eval_time = eval_time) %>%
      dplyr::filter(!is.na(observed))
  }
  
  get_all_cal_data <- function(eval_time) {
    purrr::map_dfr(eval_time, get_each_cal_data)
  }
  
  cal_plot_data <- get_all_cal_data(time_points)
  
  save(cal_plot_data, file = "cal_plot_data.RData")
} else {
  load("cal_plot_data.RData")
}
```

For the NY building complaint data, let’s look at evaluation times of 10 and 100 days. First, we can examine the distribution of the probability predictions at both time points: 

```{r}
#| label: cal-hist
#| echo: false
#| warning: false
binary_encoding %>% 
  dplyr::filter(.eval_time %in% c(10, 100) & !is.na(obs_class)) %>%
  ggplot(aes(x = 1 - .pred_survival, weight = .weight_censored)) + 
  geom_vline(xintercept = 1 / 2, col = "blue", lty = 2) +
  geom_histogram(col = "white", bins = 20) + 
  facet_grid(obs_class ~ .eval_time) +
  lims(x = 0:1) +
  labs(x = "probability of event at or before eval time") +
  theme_bw()
```

> The range of probabilities at 10 days is almost the entire range, and there is moderate separation between the two. However, at 100 days, the smallest probability prediction is `r round(small_100, 2)`. Clearly, these predictions are not well calibrated. The calibration plots tell the same story:

Reviewer: We said ealier (line 176) that the Brier score, i.e., the calibration, improved with evaluation time. It looks pretty good at eval_time = 100 days. This here contradicts that. 🤔

Author: I've clarified some things and will request rereview
